model/LTIDeliveryTool.php

```diff
@@ -68,34 +71,15 @@ public function linkDeliveryExecution(core_kernel_classes_Resource $link, $userU
         return !is_null($link);
     }
 
-    public function getFinishUrl(LtiMessage $ltiMessage = null, DeliveryExecution $deliveryExecution = null)
+    /**
+     * @param DeliveryExecution|null $deliveryExecution
+     * @return mixed
+     */
+    public function getFinishUrl(DeliveryExecution $deliveryExecution = null)
     {
-        $session = \common_session_SessionManager::getSession();
-        /** @var LtiLaunchData $launchData */
-        $launchData = $session->getLaunchData();
-        if (
-            $launchData->hasVariable(DeliveryTool::PARAM_SKIP_THANKYOU) && $launchData->getVariable(DeliveryTool::PARAM_SKIP_THANKYOU) == 'true'
-            && $launchData->hasReturnUrl()
-        ) {
-            $redirectUrl = $launchData->getReturnUrl();
-        } else {
-            $redirectUrl = _url('thankYou', 'DeliveryRunner', 'ltiDeliveryProvider');
-        }
+        $ltiNavigationService = $this->getServiceLocator()->get(LtiNavigationService::SERVICE_ID);
 
-        if ($deliveryExecution !== null) {
-            $urlParts = parse_url($redirectUrl);
-            if (!isset($urlParts['query'])) {
-                $urlParts['query'] = '';
-            }
-            parse_str($urlParts['query'], $params);
-            if ($ltiMessage) {
-                $params = array_merge($params, $ltiMessage->getUrlParams());
-            }
-            $params['deliveryExecution'] = $deliveryExecution->getIdentifier();
-            $urlParts['query'] = http_build_query($params);
-            $redirectUrl = $urlParts['scheme'] . '://' . $urlParts['host'] . $urlParts['path'] . '?' . $urlParts['query'];
-        }
-        return $redirectUrl;
+        return $ltiNavigationService->getReturnUrl($this->getLtiLaunchData(), $deliveryExecution);
```

Comment:
Good catch. I think we can make this parameter required. 
In current implementation it doesn't work with $deliveryExecution = null anyway because it's used to prepare LtiMessage:
ltiDeliveryProvider - https://github.com/oat-sa/extension-tao-ltideliveryprovider/blob/master/controller/DeliveryRunner.php#L176
ltiProctoring - https://github.com/oat-sa/extension-lti-proctoring/blob/master/controller/DeliveryServer.php#L79

Reply:
Updated